modules/internal/localeModels/gpu/LocaleModel.chpl

```diff
@@ -0,0 +1,345 @@
+
+    iter getChildIndices() : int {
+      for idx in {0..#numSublocales} do // chpl_emptyLocaleSpace do
+        yield idx;
+    }
+
+    override proc getChild(idx:int) : locale {
+      if boundsChecking then
+        if (idx < 0) || (idx >= numSublocales) then
+          halt("sublocale child index out of bounds (",idx,")");
+      return new locale(childLocales[idx]);
+    }
+
+    iter getChildren() : locale  {
+      for loc in childLocales do
+        yield new locale(loc);
+    }
+
+    proc getChildArray() {
+      return [loc in childLocales] loc;
+    }
+
+    //------------------------------------------------------------------------{
+    //- Implementation (private)
+    //-
+    proc setup() {
+      helpSetupLocaleGPU(this, numSublocales, local_name, CPULocale, GPULocale);
+    }
+    //------------------------------------------------------------------------}
+  }
+
+  //
+  // An instance of this class is the default contents 'rootLocale'.
+  //
+  // In the current implementation a platform-specific architectural description
+  // may overwrite this instance or any of its children to establish a more customized
+  // representation of the system resources.
+  //
+  class RootLocale : AbstractRootLocale {
+
+    const myLocaleSpace: domain(1) = {0..numLocales-1};
+    pragma "unsafe"
+    var myLocales: [myLocaleSpace] locale;
+
+    proc init() {
+      super.init(nilLocale);
+      nPUsPhysAcc = 0;
+      nPUsPhysAll = 0;
+      nPUsLogAcc = 0;
+      nPUsLogAll = 0;
+      maxTaskPar = 0;
+    }
+
+    // The setup() function must use chpl_initOnLocales() to iterate (in
+    // parallel) over the locales to set up the LocaleModel object.
+    // In addition, the initial 'here' must be set.
+    proc setup() {
+      helpSetupRootLocaleNUMA(this);
+    }
+
+    // Has to be globally unique and not equal to a node ID.
+    // We return numLocales for now, since we expect nodes to be
+    // numbered less than this.
+    // -1 is used in the abstract locale class to specify an invalid node ID.
+    override proc chpl_id() return numLocales;
+    override proc chpl_localeid() {
+      return chpl_buildLocaleID(numLocales:chpl_nodeID_t, c_sublocid_none);
+    }
+    override proc chpl_name() return local_name();
+    proc local_name() return "rootLocale";
+
+    override proc writeThis(f) throws {
+      f <~> name;
+    }
+
+    override proc getChildCount() return this.myLocaleSpace.size;
+
+    proc getChildSpace() return this.myLocaleSpace;
+
+    iter getChildIndices() : int {
+      for idx in this.myLocaleSpace do
+        yield idx;
+    }
+
+    override proc getChild(idx:int) return this.myLocales[idx];
+
+    iter getChildren() : locale  {
+      for loc in this.myLocales do
+        yield loc;
+    }
+
+    override proc getDefaultLocaleSpace() const ref return this.myLocaleSpace;
+    override proc getDefaultLocaleArray() const ref return myLocales;
+
+    override proc localeIDtoLocale(id : chpl_localeID_t) {
```

Comment:
I think this one should probably look like the same-named function from the 'apu' locale model, where the locale object returned is either that of the top-level locale or a sublocale, depending on whether there's a specific sublocale number encoded in the localeID passed into it.